python/lsst/daf/butler/core/storedFileInfo.py

```diff
@@ -21,25 +21,70 @@
 
 from __future__ import annotations
 
-__all__ = ("StoredFileInfo", "StoredDatastoreItemInfo")
+__all__ = ("SerializedStoredDatastoreItemInfo", "StoredFileInfo", "StoredDatastoreItemInfo")
 
+import functools
 import inspect
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Dict, Optional, Type
 
 from lsst.resources import ResourcePath
+from lsst.utils import doImportType
+from lsst.utils.introspection import get_full_type_name
+from pydantic import BaseModel
 
 from .formatter import Formatter, FormatterParameter
 from .location import Location, LocationFactory
 from .storageClass import StorageClass, StorageClassFactory
 
 if TYPE_CHECKING:
+    from ..registry import Registry
     from .datasets import DatasetId, DatasetRef
+    from .dimensions import DimensionUniverse
 
 # String to use when a Python None is encountered
 NULLSTR = "__NULL_STRING__"
 
 
+class SerializedStoredDatastoreItemInfo(BaseModel):
+    """Representation of `StoredDatastoreItemInfo` used for serialization.
+
+    Note that this class supports polymorphism by storing actual class name of
+    the items together with all other data members. It depends on a correct
+    implementation of ``to_record`` and ``from_record`` methods in each
+    sub-class.
+    """
+
+    class_name: str
+    """Actual fully-qualified name of the item class."""
```

Comment:
This may be more complicated (if done correctly). The `SerializedDatastoreRecordData.records` is now indexed by the table name, and potentially the same table name could have different schema and record type for different datastores. We now have only FileDatastore, so this may not matter, but if we ever to add a different datastore type it may be worth thinking about how to avoid future complications. I guess one way is to solve this to do something like:

    records: Dict[str, Dict[str, List[Dict[str, Any]]]

or 

    records: Dict[Tuple[str, str], List[Dict[str, Any]]


with the first index being the table name and second is the record class name. But I also need to think about your proposed change (on pipe_base PR) to index it by dataset ID instead, I'll figure something out.

Reply:
Another option that looked attractive but had other complications is to let datastore handle deserialization. Record type is specific to datastore so it's a natural place to do that. Complication is that I would need to get an access to a datastore by its name, which is probably not trivial.

Reply:
The name is a public property of an instantiated datastore. `butler.datastore` is all you have but there doesn't seem to be a problem with passing records to a method of `butler.datastore` and then allowing the datastore to handle it. It knows its own name and a chained datastore can iterate over each datastore in the chain and only pass records that have a matching name.

Reply:
Unfortunately it's harder than I hoped, `butler.datastore` is not currently accessible anywhere in the chain of `from_simple()` calls. QuantumGraph loading in general does not require butler to even be present today, and I'm reluctant to add it as an (optional) dependency for `load` method. Another approach to try is a sort of singleton registry for deserializers.

Reply:
I should have read the code that was being discussed. No, adding datastore to from_simple seems like a bad idea. I thought we were talking about something downstream of from_simple that would be handed to a chained datastore (but I imagine that was the code you wrote a couple of weeks back).